### ISSUE TITLE
Provide API to check color contrast ratio and debug them for components on colored surfaces

### ISSUE THREAD

**Issue by pylapp** [Orange-OpenSource/ouds-ios#656]
# Context

Today, we have several profiles of users:
- junior developers with maybe big turnovers of teams
- experimented developers (i.e. not yet seniosr)
- senior developers (i.e. skilled and serious enough)

These developers work on projects where:
- designers can be missing
- designers can be available without enough skills or not skills at all about a11y (true story)
- designers are aware of a11Y

In addition, projects can be:
- for countries where a11Y is a big issue (i.e. in some MEA countries)
- for countries where a11Y is a legal hot topic (i.e. in France with RGAA)
- for countries where a11Y is not a legal or marketing topic (true story)

These, in conclusion we can have:
- projects with designers not aware of a11Y in countries where law are strong about that and heavy fine can be decided ; or for countries with more a11Y needs than others
- projects without designers in countries where law are strong about that and heavy fine can be decided ; or for countries with more a11Y needs than others
- projects with developers not aware enough of a11Y

In addition, it is possible to provide themes in OUDS where some combinations of colors does not match WCAG 2.1 3:1 or 4.5:1 ratios, like * kof kof* the current v0.14.0 of OUDS with v0.11.0 of tokens lib).

# Need

We can provide API for users (here developers) so help to find cases where there are contrast ratios troubles.
We can also update components defined to be used on colored surface to notify users there are such contrast ratios troubles.

# Proposals

## 1 - Logs

We can imagine to display logs in DEBUG mode

## 2 - Provide API

Provide API for users if they want to test ratios.

## 3 - Unit tests

We can imagine to write unit tests for all combinations of colors ; but the number of tests will be high, and if a theme in OUDS does not amtch the requirements it will imply to have failing tests sio failing CI/CD chains always until the theme specifications are updated (good luck).

# Choosen solution(s)

Solutions 1 and 2

**Comment by pylapp** [comment]
@pya35 Normalement, maintenant, les utilisateurs pourront avoir dans les logs de leur Xcode des warnings si des couleurs pour le bouton (_OUDSButton_) et le link (_OUDSLink_) sont utilisées sur des surfaces colorées (_OUDSColoredSurface_) sans respecter les ratios 3:1 et 4.5:1 WCAG 2.1

Seuls ces deux types de composants à date sont testés car il n'y a qu'un à utiiser sur des surfaces avec un ensemble finid e couleurs prédéfinies.

Les développeurs pourront toujours vérifier les ratios pour eux s'ils veulent.

Du coup, si tu débug le code de l'app via Xcode, surveille tes logs.
Si tu as envie d'essayer, quelques infos ici : https://github.com/Orange-OpenSource/ouds-ios/wiki/41-%E2%80%90-Test-the-builds#on-simulator

Pour un premier jet ça me parait convenable, même si j'ai parfois l'impression que les calculs avec des nombres à virgule flottante donnent des imprécisions ; ceci dit quand bien même j'ai parfois trouvé plusieurs valeurs possibles selon les outils, valeurs divergeant assez peu, la conclusion est la même : il y a un paquet de cas où les ratios 3:1 et 4.5:1 ne sont pas respectés.

**Comment by pylapp** [comment]
@pya35 Pour info j'ai fait une carte consignant les logs qui sont sortis quand des ratios WCAG 2.1 n'étaient pas respectés : https://github.com/Orange-OpenSource/ouds-ios/issues/658

Reste à savoir si, pour le chevron du _link_ on doit considérer le 3:1 ou le 4.5:1, et même chose pour le bouton par exemple.

**Comment by pya35** [comment]
> [@pya35](https://github.com/pya35) Pour info j'ai fait une carte consignant les logs qui sont sortis quand des ratios WCAG 2.1 n'étaient pas respectés : [#658](https://github.com/Orange-OpenSource/ouds-ios/issues/658)
> 
> Reste à savoir si, pour le chevron du _link_ on doit considérer le 3:1 ou le 4.5:1, et même chose pour le bouton par exemple.

Contraste min. de éléments textuels 4.5:1 / non-textuels 3:1

**Comment by pylapp** [comment]
Ok, merci @pya35 !
Est-ce que tu as (j'anticipe) des références de specs mentionnant le 7:1 ?

**Comment by pya35** [comment]
> Ok, merci [@pya35](https://github.com/pya35) ! Est-ce que tu as (j'anticipe) des références de specs mentionnant le 7:1 ?

@pylapp Il n'y aucune obligation légale ni recommandation officielle coté WCAG concernant l'atteinte du niveau AAA. Les référentiels WCAG transposés aux mobiles suivent d'ailleurs les recos Web, à savoir le niveau AA. Mais ce niveau est le **strict minimum** attendu.

Par contre notre propre Design System recommande **fortement** d'augmenter le contraste pour les apps mobiles :  https://system.design.orange.com/0c1af118d/p/2531b9-accessibility-compliance

Une autre approche est d'accepter le niveau AA par défaut mais que le niveau AAA soit atteint en activant le mode contraste élevé sur les mobiles, C'est ce qu'il serait peut-être pertinent de faire si le niveau AAA n'est pas atteignable par défaut.

**Comment by B3nz01d** [comment]
Completed during Sprint 18
